multicluster/controllers/multicluster/stale_controller.go

```diff
@@ -0,0 +1,199 @@
+	}
+	return reconciler
+}
+
+//+kubebuilder:rbac:groups="",resources=services,verbs=get;list;watch;delete
+//+kubebuilder:rbac:groups=multicluster.x-k8s.io,resources=serviceimports,verbs=get;list;watch;delete
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=resourceexports,verbs=get;list;watch
+
+func (c *StaleController) cleanup() error {
+	if *c.remoteCommonAreaManager == nil {
+		return errors.New("ClusterSet has not been initialized properly, no available remote common area")
+	}
+
+	remoteCluster, err := getRemoteCommonArea(c.remoteCommonAreaManager)
+	if err != nil {
+		return err
+	}
+
+	svcImpList := &k8smcsv1alpha1.ServiceImportList{}
+	if err := c.List(ctx, svcImpList, &client.ListOptions{}); err != nil {
+		return err
+	}
+
+	svcList := &corev1.ServiceList{}
+	if err := c.List(ctx, svcList, &client.ListOptions{}); err != nil {
+		return err
+	}
+
+	resImpList := &mcsv1alpha1.ResourceImportList{}
+	if err := remoteCluster.List(ctx, resImpList, &client.ListOptions{Namespace: remoteCluster.GetNamespace()}); err != nil {
+		return err
+	}
+
+	svcImpItems := svcImpList.Items
+	var mcsSvcItems []corev1.Service
+	for _, svc := range svcList.Items {
+		if _, ok := svc.Annotations[common.AntreaMCServiceAnnotation]; ok {
+			mcsSvcItems = append(mcsSvcItems, svc)
+		}
+	}
+
+	for _, resImp := range resImpList.Items {
+		for k, svc := range mcsSvcItems {
+			if svc.Name == common.AntreaMCSPrefix+resImp.Spec.Name && svc.Namespace == resImp.Spec.Namespace {
+				// Set the valid Service item as empty Service, then all left non-empty items should be removed.
+				mcsSvcItems[k] = corev1.Service{}
+			}
+		}
+
+		for n, svcImp := range svcImpItems {
+			if svcImp.Name == resImp.Spec.Name && svcImp.Namespace == resImp.Spec.Namespace {
+				svcImpItems[n] = k8smcsv1alpha1.ServiceImport{}
+			}
+		}
+	}
+
+	for _, svc := range mcsSvcItems {
+		s := svc
+		if s.Name != "" {
+			klog.InfoS("clean up Service", "service", klog.KObj(&s))
+			if err := c.Client.Delete(ctx, &s, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+
+	for _, svcImp := range svcImpItems {
+		si := svcImp
+		if si.Name != "" {
+			klog.InfoS("clean up ServiceImport", "serviceimport", klog.KObj(&si))
+			if err = c.Client.Delete(ctx, &si, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Enqueue will be called after StaleController is initialized.
+func (c *StaleController) Enqueue() {
+	// The key can be anything as we only have single item.
+	c.queue.Add("key")
+}
+
+// Run starts the StaleController and blocks until stopCh is closed.
+func (c *StaleController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting StaleController")
+	defer klog.InfoS("Shutting down StaleController")
+
+	if err := c.RunOnce(); err != nil {
```

Comment:
But we should do clean up before other controllers start to run, no?

Reply:
we only clean up those stale resources, eg, remove MC Service and ServiceImport when corresponding ResourceImport does not existing in leader cluster anymore, so it should be fine to no run before other controllers because other controllers will do nothing for a deleted resource which is already gone before reconciling and no delete event at all.

Reply:
I am thinking about any possible race condition here, e.g. stalecontroller (as it is not atomic to get imports, services, and delete) deletes new Services?

Reply:
stale controller will get ServiceList from local first, then ResourceImportList from leader, so it means if new derived Service being created but not in ServiceList, we either have new ResourceImport in ResourceImportList or not have it, in first case Service won't be deleted because there is corresponding ResourceImport, in second case, nothing will be done.